static/app/components/dataExport.tsx

```diff
@@ -17,59 +18,47 @@ export enum ExportQueryType {
   Discover = 'Discover',
 }
 
-type DataExportPayload = {
+interface DataExportPayload {
   queryType: ExportQueryType;
   queryInfo: any; // TODO(ts): Formalize different possible payloads
-};
+}
 
-type Props = {
+interface DataExportProps {
   api: Client;
   disabled?: boolean;
   organization: Organization;
   payload: DataExportPayload;
   icon?: React.ReactNode;
-};
-
-type State = {
-  inProgress: boolean;
-};
+  children?: React.ReactNode;
+}
 
-class DataExport extends React.Component<Props, State> {
-  state = this.initialState;
+function DataExport({
+  api,
+  children,
+  disabled,
+  organization,
+  payload,
+  icon,
+}: DataExportProps): React.ReactElement {
+  const previousPayload = usePrevious(payload);
+  const [inProgress, setInProgress] = React.useState(false);
 
-  componentDidUpdate({payload: prevPayload}) {
-    const {payload} = this.props;
-    if (!isEqual(prevPayload, payload)) {
-      this.resetState();
+  React.useEffect(() => {
+    if (!isEqual(payload, previousPayload)) {
+      setInProgress(false);
     }
-  }
-
-  get initialState(): State {
-    return {
-      inProgress: false,
-    };
-  }
+  }, [payload]);
 
-  resetState = () => {
-    this.setState(this.initialState);
-  };
-
-  startDataExport = () => {
-    const {
-      api,
-      organization: {slug},
-      payload: {queryType, queryInfo},
-    } = this.props;
-
-    this.setState({inProgress: true});
+  const handleDataExport = React.useCallback(() => {
+    setInProgress(true);
 
     api
-      .requestPromise(`/organizations/${slug}/data-export/`, {
+      .requestPromise(`/organizations/${organization.slug}/data-export/`, {
```

Comment:
@leeandher  I guess this is fine, but it is prone to a race condition. E.g. if user clicks on the button and a slow API call is made, then props are updated and if they then click again, we'll fire the second call and if this call resolves before the first, we'll end up with the wrong state when the first call resolves. Idk if this is a problem, maybe we expect the endpoint to always be ⚡

Reply:
@JonasBa I think it should be fine, though it's been a while since I touched this code. I think if the payload from props changes we still want to fire the second request since its a new set of data to export. 

Even if the second request resolves before the first, both would then trigger success messages and return (which is the expected behavior). The synchronous parts of the callback both set `inProgress` so they should disable repeat clicks on identical exports as well (+ they're debounced).

All that to say, this looks totally fine to me, but I might just be misunderstanding the race condition here 🤣

Reply:
Ah, yes. We still want to display the message in that case, I thought we wouldn't want that!